Add a clj-kondo hook for bad single-quotes in i18n

johnswanson opened the pull request:
For i18n, we had a lot of things like

```
(tru "this user's string will be broken")
```

During i18n, we use `java.text.MessageFormat` to format messages. These have a great idea: you can format messages with parameters like `{0}`, or escape these cases with single-quotes like `'{0}'` when you actually want the string "{0}". And of course, you can escape a single-quote itself just by doubling it, so "''" becomes a "'".

All this makes sense. But then they had a problem. What if there's an unescaped single-quote in the argument to MessageFormat?

A lesser program might have thrown an exception here, but not MessageFormat! Nay! We must persist, lads, not give up at the first sign of trouble!

Instead, let's just decide that the programmer probably *meant* to close that single-quote! So, for example, "This user's message is totally frickin broken" is interpreted AS IF you'd written "This user's message is totally frickin broken'" (which ends up as "This users message is totally frickin broken"), which - doubtless - is what you intended.

Unfortunately there are *some* few cases where - IT TURNS OUT - we don't actually want this behavior! For example,

```
(deferred-tru "Key to retrieve the JWT user's email address.")
```

becomes "Key to retrieve the JWT users email address." Which is slightly annoying.

So: let's have a kondo rule that will enforce statically that you don't have unescaped single-quotes in your template strings for `tru`, `deferred-tru`, `trs`, or `deferred-trs`.

johnswanson committed:
Add a clj-kondo hook for bad single-quotes in i18n

For i18n, we had a lot of things like

```
(tru "this user's string will be broken")
```

During i18n, we use `java.text.MessageFormat` to format messages. These
have a great idea: you can format messages with parameters like `{0}`, or
escape these cases with single-quotes like `'{0}'` when you actually
want the string "{0}". And of course, you can escape a single-quote
itself just by doubling it, so "''" becomes a "'".

All this makes sense. But then they had a problem. What if there's an
unescaped single-quote in the argument to MessageFormat?

A lesser program might have thrown an exception here, but not
MessageFormat! Nay! We must persist, lads, not give up at the first sign
of trouble!

Instead, let's just decide that the programmer probably *meant* to close
that single-quote! So, for example, "This user's message is totally
frickin broken" is interpreted AS IF you'd written "This user's message
is totally frickin broken'" (which ends up as "This users message is
totally frickin broken"), which - doubtless - is what you intended.

Unfortunately there are *some* few cases where - IT TURNS OUT - we don't
actually want this behavior! For example,

```
(deferred-tru "Key to retrieve the JWT user's email address.")
```

becomes "Key to retrieve the JWT users email address." Which is slightly
annoying.

So: let's have a kondo rule that will enforce statically that you don't
have unescaped single-quotes in your template strings for `tru`,
`deferred-tru`, `trs`, or `deferred-trs`.

rafpaf commented:
🙏🙏🙏

crisptrutski commented on .clj-kondo/src/hooks/metabase/util/i18n.clj:
```diff
@@ -0,0 +1,59 @@
+              ;; Technically a string like "foo bar's happy but i'm not" is totally fine, but we'll say it's invalid
+              ;; because... that inner string literal is probably not what we actually want. So since we've never done
+              ;; this intentionally as far as I can tell, let's just remove the possibility to avoid ambiguity.
+            (and (next chars) (= (second chars) \{))
```
Why is the `(next chars)` check required here? The second predicate is still false for single element seqs.

crisptrutski commented on .clj-kondo/src/hooks/metabase/util/i18n.clj:
```diff
@@ -0,0 +1,59 @@
+              ;; because... that inner string literal is probably not what we actually want. So since we've never done
+              ;; this intentionally as far as I can tell, let's just remove the possibility to avoid ambiguity.
+            (and (next chars) (= (second chars) \{))
+            (let [rest-str (apply str (rest chars)) ; From '{
+                  closing-index (str/index-of rest-str "'")]
```
nit: Rather than building this string you could take memory and double iteration by doing something like:

```clojure
(if-let [after-apostrophe (next (drop-while (comp not #{\'}) (rest chars)))]
  ...)
```

crisptrutski commented on .clj-kondo/src/hooks/metabase/util/i18n.clj:
```diff
@@ -0,0 +1,59 @@
+        ;; strings into shorter lines)
+        format-str (if (api/string-node? format-str-arg)
+                     (api/sexpr format-str-arg)
+                     ;; (str "a" "b" "c") => a list-node with children "str", "a", "b", and "c"
+                     (apply str (map api/sexpr (rest (:children format-str-arg)))))]
```
I'd feel a bit more secure if we validated that the first child is indeed `'str` 😬 

Doesn't seem out of the question that we'd call some custom string building function. We would probably want to fail open in that case.

johnswanson replied:
So we're already validating that the only possible first arg to `tru` or `trs` is `(str "some" "strings") or a string literal in `metabase.util.i18n` - but you're right, I shouldn't rely on that here! I'll add a check here too.

crisptrutski reviewed (approved):
The value is clearly borne out in all these broken examples it found!

johnswanson committed:
CR feedback response

- more efficient linter

- add linter for "it's a string or call to str"

- remove unnecessary `(next chars)` check

- also, unrelated, fix yet another broken test

johnswanson committed:
GDI so many tests dependent on the broken strings
